state/indexer/block/kv/kv_test.go

```diff
@@ -0,0 +1,141 @@
+package kv_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/pubsub/query"
+	blockidxkv "github.com/tendermint/tendermint/state/indexer/block/kv"
+	"github.com/tendermint/tendermint/types"
+	db "github.com/tendermint/tm-db"
+)
+
+func TestBlockIndexer(t *testing.T) {
+	store := db.NewPrefixDB(db.NewMemDB(), []byte("block_events"))
+	indexer := blockidxkv.New(store)
+
+	require.NoError(t, indexer.Index(types.EventDataNewBlockHeader{
+		Header: types.Header{Height: 1},
+		ResultBeginBlock: abci.ResponseBeginBlock{
+			Events: []abci.Event{
+				{
+					Type: "begin_event",
+					Attributes: []abci.EventAttribute{
+						{
+							Key:   []byte("proposer"),
+							Value: []byte("FCAA001"),
+							Index: true,
+						},
+					},
+				},
+			},
+		},
+		ResultEndBlock: abci.ResponseEndBlock{
+			Events: []abci.Event{
+				{
+					Type: "end_event",
+					Attributes: []abci.EventAttribute{
+						{
+							Key:   []byte("foo"),
+							Value: []byte("100"),
+							Index: true,
+						},
+					},
+				},
+			},
+		},
+	}))
+
+	for i := 2; i < 12; i++ {
+		var index bool
+		if i%2 == 0 {
+			index = true
+		}
+
+		require.NoError(t, indexer.Index(types.EventDataNewBlockHeader{
+			Header: types.Header{Height: int64(i)},
+			ResultBeginBlock: abci.ResponseBeginBlock{
+				Events: []abci.Event{
+					{
+						Type: "begin_event",
+						Attributes: []abci.EventAttribute{
+							{
+								Key:   []byte("proposer"),
+								Value: []byte("FCAA001"),
+								Index: true,
+							},
+						},
+					},
+				},
+			},
+			ResultEndBlock: abci.ResponseEndBlock{
+				Events: []abci.Event{
+					{
+						Type: "end_event",
+						Attributes: []abci.EventAttribute{
+							{
+								Key:   []byte("foo"),
+								Value: []byte(fmt.Sprintf("%d", i)),
+								Index: index,
+							},
+						},
+					},
+				},
+			},
+		}))
+	}
+
+	testCases := map[string]struct {
```

Comment:
Test cases look good. The standard types are:

* `block.height = X` for begin and end
* `begin_event.proposer` for begin
* `end_event.val_updates` (how do you properly query this?)

And when searching for eg. transfers in begin and end block, I could do: `transfer.sender = 'ABC123'` and it should match both begin and end block events?

If I understand this correctly, then this looks great

Reply:
Yes, that's correct!